design/anonymousTokens.md

```diff
@@ -0,0 +1,96 @@
+        make that change (because they all expect a token).
+
+## Document purpose
+
+Propose a preliminary design for KBase auth service support of anonymous tokens (ATs)
+that allow anonymous users to have a transient identity when interacting with KBase. Discuss how
+those tokens would interact with the rest of the KBase infrastructure and what changes would be
+necessary.
+
+## Anonymous tokens
+
+* The KBase auth service would be enhanced to offer AT support.
+* ATs would have the same fields as regular tokens except that the the `type` would
+  always be `Anonymous`.
+* ATs would have a random user name assigned upon creation.
+  * Could be a concatenated dictionary style name for mnemonic purposes
+    (e.g. [https://gfycat.com/shoddydismalamoeba](https://gfycat.com/shoddydismalamoeba)).
+      * Need to check collision frequency.
+* ATs would exist in a separate token and user namespace from standard tokens.
+* ATs could not be used to create other tokens.
+* ATs would be prefixed with a string to indicate they are anonymous
+  (e.g. `***ANONYMOUS***-[remainder of token goes here]`)
+* ATs would be created, introspected, and revoked with separate APIs from normal tokens.
+  * Token creation requires no authentication.
+* The administration UI would have AT and user lookup functionality and the ability
+  to revoke said tokens.
+* The administration UI would have the ability to set the lifetime of ATs. The default lifetime
+  could be fairly long since ATs grant few privileges.
+* All endpoints except for the AT specific endpoints would respond with an error when
+  presented with an AT.
+  
+## Interactions with KBase infrastructure
+
+* Most KBase core services (including data services) should never see an an AT.
+* Narrative:
+  * When a user without a KBase cookie in the browser loads the KBase UI, an anonymous token would
+    be generated and stored in a cookie (potentially the same cookie).
+  * Starting a narrative would assign the docker container to the username in the AT.
+    * Narratives started in this manner would always be read-only to the user and must
+      be public.
+    * The Narrative backend would not pass the AT on to any services.
+  * The UI should show that the user is in anonymous mode, their transient username, and allow
+    them to log in normally.
+* Downloads
+  * Note that DLs are significantly more work but are orthogonal to the Narrative work.
+  * JSON DLs should work as is as long as the front end does not pass the AT to DIE.
+  * Other downloads
+    * The NJS would need to understand anonymous tokens (to allow for job
+      tracking purposes) and have the ability to start jobs without a token.
+      * The NJS or catalog could have a list of apps that may be started without a
+        token - these would presumably be DL apps only.
+    * The SDK would have to support starting apps without a token.
+    * DL apps would have to support running without a token.
```

Comment:
And the Narrative should handle starting and tracking without a token

Reply:
Clarified that the narrative needs to understand and use ATs, since it needs to pass them on to NJS and SS. 

Also I rewrote the DL apps stuff since they're going to need a token to know where to stick data in the staging area.